### PR TITLE
gh-107932: Fix dis module for bytecode that does not have an associated source line

### LINKED ISSUE #107932: `dis.dis()` reports line numbers where there are none.
Take this example function from https://github.com/python/cpython/issues/107901:
```
def f():
   for i in range(30000000):
      if not i%1000000:
          pass
```
`dis.dis` produces the following output:
```
  1           0 RESUME                   0

  2           2 LOAD_GLOBAL              1 (range + NULL)
             12 LOAD_CONST               1 (30000000)
             14 CALL                     1
             22 GET_ITER
        >>   24 FOR_ITER                14 (to 56)
             28 STORE_FAST               0 (i)

  3          30 LOAD_FAST                0 (i)
             32 LOAD_CONST               2 (1000000)
             34 BINARY_OP                6 (%)
             38 TO_BOOL
             46 POP_JUMP_IF_FALSE        2 (to 52)
             48 JUMP_BACKWARD           14 (to 24)

  4     >>   52 JUMP_BACKWARD           16 (to 24)

  2     >>   56 END_FOR
             58 RETURN_CONST             0 (None)
```
Which has incorrect line numbers.
The issue is not that the line numbers are wrong, but that you can't tell from the `dis` output.
The whole point of `dis` is show what is going on at the bytecode level, so it is failing if it gives wrong line numbers.
The actual line numbers can be see from `co_lines()`
```
>>> list(f.__code__.co_lines())
[(0, 2, 1), (2, 30, 2), (30, 48, 3), (48, 52, None), (52, 56, 4), (56, 60, 2)]
```

The correct output should be:
```
   1           0 RESUME                   0

   2           2 LOAD_GLOBAL              1 (range + NULL)
              12 LOAD_CONST               1 (30000000)
              14 CALL                     1
              22 GET_ITER
         >>   24 FOR_ITER                14 (to 56)
              28 STORE_FAST               0 (i)

   3          30 LOAD_FAST                0 (i)
              32 LOAD_CONST               2 (1000000)
              34 BINARY_OP                6 (%)
              38 TO_BOOL
              46 POP_JUMP_IF_FALSE        2 (to 52)
None          48 JUMP_BACKWARD           14 (to 24)

   4     >>   52 JUMP_BACKWARD           16 (to 24)

   2     >>   56 END_FOR
              58 RETURN_CONST             0 (None)
```

### Linked PRs
* gh-107988
* gh-108479
* gh-108478

### PR DESCRIPTION
**Pull request by CorvinM**
Fix dis module for bytecode that does not have an associated source line.
fixes #107932 

This does change behavior of many of the dis.py public functions to display lines with no source line mapping as such.
At the very least it changes:
findlinestarts - now returns starts of bytecode blocks that dont have a source line with lineno=None
dis/disassemble/distb/disco - now separates and labels lineno=None bytecode blocks, previously it lumped them into the previous block (see #107932)
get_instructions & Instruction - Instruction objects now have starts_line=False to indicate that the instruction does not start a line.

* Issue: gh-107932

### CODE REVIEW

File: Lib/dis.py [hunk 1]
```diff
@@ -376,8 +376,9 @@ def _disassemble(self, lineno_width=3, mark_as_current=False, offset_width=4):
         fields = []
         # Column: Source code line number
         if lineno_width:
-            if self.starts_line is not None:
-                lineno_fmt = "%%%dd" % lineno_width
+            if self.starts_line is not False:
```
Comment (markshannon) [inline Lib/dis.py]: Rather than overload `starts_line` in this way, maybe split into two attributes, `starts_line` which is just a boolean and `line_number` which is the actual line number.
The logic is then clearer:
```Py
if self.starts_line:
    lineno_fmt = "%%%dd" if self.line_number is not None else "%%%ds"
    lineno_fmt = lineno_fmt % lineno_width
    fields.append(lineno_fmt % self.line_number)
else:
    fields.append(' ' * lineno_width)
```
```

**Review by markshannon — CHANGES_REQUESTED** [review]
Thanks for working on this. 
The changes look correct, but the use of `False` as a special value for `starts_line` is a bit hard to follow.

`Lib/test/test_dis.py` is an endless source of merge conflicts, so I'll try merge ASAP after you've made the changes.

**Comment by CorvinM** [comment]
@markshannon Thanks for the quick review! I have made the requested changes; please review again. I have also updated the docs to reflect the changes and resolved the merge conflict (for now). I've listed the changes in the docs as happening in 3.13 (is this the correct procedure for the project?)

**Review by markshannon — APPROVED** [review]
Thanks, looks good.

**Comment by erlend-aasland** [comment]
`main` has been broken since this was merged.

**Comment by markshannon** [comment]
https://github.com/python/cpython/pull/108478
